crates/core/src/metadata_store/providers/objstore/optimistic_store.rs

```diff
@@ -0,0 +1,254 @@
+        }
+        Ok(())
+    }
+
+    async fn get_or_update_store(&self) -> anyhow::Result<Arc<Option<CachedStore>>> {
+        self.maybe_update_cached_store(false).await?;
+        let maybe_current_store = self.latest_store_cache.load_full();
+        Ok(maybe_current_store)
+    }
+
+    async fn compare_and_swap(
+        &self,
+        mut op: impl FnMut(&ImmutableStore) -> Result<ImmutableStore, WriteError>,
+    ) -> Result<(), WriteError> {
+        for _retry_count in 0..25 {
+            let current_store = self
+                .get_or_update_store()
+                .await
+                .map_err(|e| WriteError::Network(e.into()))?;
+            let res = match current_store.as_ref() {
+                Some(cached_store) => op(&cached_store.store),
+                None => unreachable!(),
+            };
+            let Ok(new_store) = res else {
+                // a precondition failure. the actual failure is reflected in the Err.
+                return res.map(|_| ());
+            };
+
+            let new_store_bytes = new_store.serialize().map_err(WriteError::Store)?;
+
+            match self
+                .version_repository
+                .try_create_version(new_store.current_version(), new_store_bytes)
+                .await
+            {
+                Ok(_) => {
+                    // All good!
+                    tracing::info!(
+                        "successfully published a new metadata store version {}",
+                        new_store.current_version()
+                    );
+                    self.latest_store_cache
+                        .store(Arc::new(Some(CachedStore::new(new_store))));
+                    return Ok(());
+                }
+                Err(VersionRepositoryError::AlreadyExists(_)) => {
+                    // refresh and try again
+                    let delay = {
+                        let mut rng = rand::thread_rng();
+                        rng.gen_range(50..300)
+                    };
+                    tokio::time::sleep(Duration::from_millis(delay)).await;
+                    self.maybe_update_cached_store(true)
+                        .await
+                        .map_err(|e| WriteError::Network(e.into()))?;
+                }
+                Err(e) => {
+                    return Err(WriteError::Store(e.into()));
+                }
+            }
+        }
+        Err(WriteError::Internal("retry count exhausted".to_string()))
+    }
+}
+
+#[async_trait::async_trait]
+impl MetadataStore for OptimisticLockingMetadataStore {
+    async fn get(&self, key: ByteString) -> Result<Option<VersionedValue>, ReadError> {
+        let cached_store = self
+            .get_or_update_store()
+            .await
+            .map_err(|e| ReadError::Network(e.into()))?;
+
+        let store = cached_store
+            .as_ref()
+            .as_ref()
+            .expect("BUG: a cached value must be stored after a call to a get_or_update_store()");
```

Comment:
Shouldn't `get_or_update_store` then return a `CachedStore` instead of `Option<CachedStore>`?